draft-howard-rats-coserv.md

```diff
@@ -225,26 +307,51 @@ The environment selector forms the main body of the query, and its CDDL is given
-
-- **Instance**: An instance is an environment that is unique to an individual and identifiable Attester, such as a single computing device or component.
-
-- **Group**: A group is a collection of common Attester instances that are collected together based on some defined semantics. For example, Attesters may be put into groups for the purpose of anonymity.
+Environments can be specified according to instance, group or class. See {{secenvironments}} for details.
 
 Although these three environment definitions are mutually-exclusive in a CoSERV query, all three support multiple entries.
-This is to gain efficiency by allowing the consumer (such as a Verifier) to query for multiple artifacts in a single transaction.
+This is to gain efficiency by allowing the consumer (Verifier) to query for multiple artifacts in a single transaction.
 For example, where artifacts are being indexed by instance, it would be possible to specify an arbitrary number of instances in a single query, and therefore obtain the artifacts for all of them in a single transaction.
 Likewise for classes and groups.
 However, it would not be possible for a single query to specify more than one kind of environment.
 For example, it would not be possible to query for both class-level and instance-level artifacts in a single CoSERV transaction.
 
+#### Selector Semantics
+
+When multiple environment selectors are present in a single query, such as multiple instances or multiple groups, the implementation of the artifact producer MUST consider these to be alternatives, and hence use a logical `OR` operation when applying the query to its internal data stores.
+
+Below is an illustrative example of how a CoSERV query for endorsed values, selecting for multiple Attester instances, might be transformed into a semantically-equivalent SQL database query:
+
+~~~sql
+SELECT *
+  FROM endorsed_values
+ WHERE ( instance-id = "At6tvu/erQ==" ) OR
+       ( instance-id = "iZl4ZVY=" )`
+~~~
+
+The same applies for class-based selectors. However, since class selectors are themselves composed of multiple inner fields, the implementation of the artifact producer MUST use a logical `AND` operation in consideration of the inner fields for each class.
+
+Also, for class-based selectors, any unset fields in the class are assumed to be wildcard (`*`), and therefore match any value.
+
+Below is an illustrative example of how a CoSERV query for reference values, selecting for multiple Attester classes, might be transformed into a semantically-equivalent SQL database query:
+
+~~~sql
+SELECT *
+  FROM reference_values
+ WHERE ( class-id = "iZl4ZVY=" AND class-vendor = "ACME Inc." ) OR
+       ( class-id = "31fb5abf-023e-4992-aa4e-95f9c1503bfa" )
+~~~
+
+### Timestamp
+
+The `timestamp` field records the date and time at which the query was made, formatted according to {{RFC3339}}.
```

Comment:
In this case, {{Section 3.4.1 of -cbor}} provides a more precise reference than {{RFC3339}}.